common/types/activation.go

```diff
@@ -299,18 +241,28 @@ func (atx *ActivationTx) MarshalLogObject(encoder log.ObjectEncoder) error {
 	return nil
 }
 
-// CalcAndSetID calculates and sets the cached ID field. This field must be set before calling the ID() method.
-func (atx *ActivationTx) CalcAndSetID() error {
+// Initialize calculates and sets the cached ID field. This field must be set before calling the ID() method.
+func (atx *ActivationTx) Initialize() error {
+	if atx.ID() != EmptyATXID {
+		return fmt.Errorf("ATX already initialized")
+	}
+
 	if atx.Signature == EmptyEdSignature {
-		return fmt.Errorf("cannot calculate ATX ID: sig is nil")
+		return fmt.Errorf("ATX must be signed before initialization")
 	}
 
 	if atx.MsgHash != BytesToHash(atx.HashInnerBytes()) {
 		return fmt.Errorf("bad message hash")
 	}
 
-	id := ATXID(CalcObjectHash32(atx))
-	atx.id = &id
+	h := hash.New()
+	if _, err := h.Write(atx.MsgHash[:]); err != nil {
+		return fmt.Errorf("failed to write to hash")
+	}
+	if _, err := scale.EncodeByteSlice(scale.NewEncoder(h), atx.Signature[:]); err != nil {
```

Comment:
i don't understand this change. i thought the only change we want to do is to exclude signature from hashing ID.

Reply:
I don't read that from #4139. The only change is to use canonical ED signatures instead of our custom library.

I know the addition of the `NodeID` field in the initial ATX was proposed to make this possible, but I would rather that field again, since we are broadcasting the Node ID for signature validation anyway and the data would be redundant.

The alternative would be to not include signature in ATX ID calculation, but should this change then also be applied to all other messages with signatures (e.g. Ballots)? Also we would then need to enforce strict signatures, which might not be supported by various implementations of ed25519.

Reply:
Regarding the actual change here: `CalcObjectHash32` encodes and hashes the atx. We already have the hash in `atx.MsgHash` and verified it already against `atx.HashInnerBytes`.

So instead I just do what `CalcObjectHash32` with the cached `atx.MsgHash`.

Reply:
> Regarding the actual change here: CalcObjectHash32 encodes and hashes the atx. We already have the hash in atx.MsgHash and verified it already against atx.HashInnerBytes.

So instead I just take the atx.MsgHash and hash it together with atx.Signature (like Ballot does).

the old ID calculation include the whole content of InnerActivationTx, ATXMetadata (not just the hash) and the Signature. we just need to remove Signature here.

Reply:
I don't understand why we have `ATXMetadata` as part of the ATX.

Target Epoch is derived from the positioning ATX and doesn't need to be included into the ATX. The MsgHash will be identical to the ATXID when we remove the signature from the ID calculation.

Shall this type be removed completely?

Reply:
> what dose strict signatures mean?

Depending on the signature scheme there might be multiple valid signatures for the same message and key.

Now that I think about it, this might not actually be an issue. We just need to check what happens if the same (valid) ATX is broadcasted twice with different (valid) signatures. Now those ATX would have different IDs and the node would be flagged as malicious, afterwards both ATX would be valid.

Reply:
maybe it's worth adding a sanity check on ID calculation so there is no ambiguity.

Reply:
ultimately i think just writing atx.MsgHash[:] is sufficient for unique ID?

Reply:
> I don't understand why we have ATXMetadata as part of the ATX.

it's there for generating malfeasance proof. the signature signs that hash and epoch. if two different msg for the same epoch appear in the network, the signer is considered malicious.

Reply:
> it's there for generating malfeasance proof. the signature signs that hash and epoch. if two different msg for the same epoch appear in the network, the signer is considered malicious.

It's also encoded and broadcasted with every ATX although the data inside can be derived by the node and doesn't necessarily need to be broadcasted.

Regarding malicious behaviour detection: If we make the hash of `InnerActivationTx` the new ID it could replace `MsgHash` I'm just not sure how to approach `TargetEpoch`. I can't find any place where this value is validated to be `atx.PubLayerID.GetEpoch() + 1`. What if a node just uses a different number here?

Reply:
I think `TargetEpoch` might need to be replaced with `atx.PubLayerID` and checked the value for the two ATXs differs by less then `layersPerEpoch`.

Reply:
> It's also encoded and broadcasted with every ATX although the data inside can be derived by the node and doesn't necessarily need to be broadcasted.

yes. Dmitry mentioned that optimization for hare too.

> I can't find any place where this value is validated to be atx.PubLayerID.GetEpoch() + 1. What if a node just uses a different number here?

true that it wasn't checked in my implementation. but target epoch is part of the signed data.

Reply:
i am changing atx to store PublishEpoch instead of PublayerID. Target epoch is always +1 to publish, so either one works for equivocation purposes

Reply:
So how shall I proceed with signature creation for ATXs? If its OK for you I would do the following:

- Scale encode `InnerActivationTx`
- Hash of encoded data -> `ATXID`
- Signature over encoded data -> `atx.Signature`

This would a) remove the signature from ID calculation and b) remove the need to have both `MsgHash` and `ATXID`. This would also exclude `ATXMetadata.TargetEpoch` from both the ID and signature calculation, since it will be replaced by @dshulyak changes in the near future anyway?

Opinions? @dshulyak @countvonzero

Reply:
for signature input should be as small as possible, that satisfies equivocation: `PublishEpoch/TargetEpoch, MsgHash`. as this is the only data that needs to be kept around if there is an actual equivocation

both PublishEpoch or TargetEpoch don't make any practical difference for equivocation

Using hash of InnerActivationTx as ATXID and MsgHash sounds good to me

Reply:
publish/target epoch does play a role in equivocation verification.

if signer only signs `MsgHash` but not `TargetEpoch`, someone could put together an atx from epoch N and one from N+X in the malfeasance proof.

i also responded in another thread that using `MsgHash` solely for ID is good.

Reply:
> both PublishEpoch or TargetEpoch don't make any practical difference for equivocation

i probably wrote it in a strange way, i meant that using any of those two for signature is ok

Reply:
i might have jumped on the last sentence. your first sentence
 
> for signature input should be as small as possible, that satisfies equivocation: PublishEpoch/TargetEpoch, MsgHash. 

is perfectly clear. sorry.